pkg/agent/plugin/nodeattestor/devid/tpm/context.go

```diff
@@ -0,0 +1,199 @@
+		// an error.
+		_, ok := c.rwc.(*tpmutil.EmulatorReadWriteCloser)
+		if ok {
+			return
+		}
+
+		err := c.rwc.Close()
+		if err != nil {
+			c.log.Warn(fmt.Sprintf("Failed to close TPM: %v", err))
+		}
+	}
+}
+
+// SolveDevIDChallenge request the TPM to sign the provided nonce using the loaded
+// DevID credentials.
+func (c *Context) SolveDevIDChallenge(nonce []byte) ([]byte, error) {
+	return c.DevID.Sign(nonce)
+}
+
+// SolveCredActivationChallenge runs credential activation on the TPM. It proves
+// that the attestation key resides on the same TPM as the endorsement key.
+func (c *Context) SolveCredActivationChallenge(credentialBlob, secret []byte) ([]byte, error) {
+	hSession, err := c.createPolicySession()
+	if err != nil {
+		return nil, err
+	}
+
+	b, err := tpm2.ActivateCredentialUsingAuth(
+		c.rwc,
+		[]tpm2.AuthCommand{
+			{Session: tpm2.HandlePasswordSession},
+			{Session: hSession},
+		},
+		c.AK.Handle,
+		c.EK.Handle(),
+		credentialBlob,
+		secret,
+	)
+	if err != nil {
+		// Flush only in case of error. If the command executes successfully it
+		// closes the session. Closing it again produces an error.
+		c.flushContext(hSession)
+		return b, fmt.Errorf("failed to activate credential: %w", err)
+	}
+
+	return b, nil
+}
+
+// LoadKey loads a key pair into the TPM.
+func (c *Context) LoadKey(pub, priv []byte) (*SigningKey, error) {
+	sk, err := LoadSigningKey(c.rwc, pub, priv, c.log)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load key on TPM: %w", err)
+	}
+	return sk, nil
+}
+
+// GetEKCert returns TPM endorsement certificate.
+func (c *Context) GetEKCert() ([]byte, error) {
+	EKCert, err := tpm2.NVRead(c.rwc, ekRSACertificateHandle)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read NV index %08x: %w", ekRSACertificateHandle, err)
+	}
+
+	return EKCert, nil
+}
+
+// RegenerateEK regenerates the Endorsement Key using the default RSA template
+func (c *Context) RegenerateEK() (*tpm2tools.Key, error) {
+	return tpm2tools.EndorsementKeyRSA(c.rwc)
+}
+
+// EncodePublicEK returns the public part of the Endorsement Key encoded in
+// TPM wire format
+func (c *Context) EncodePublicEK() ([]byte, error) {
+	publicEK, _, _, err := tpm2.ReadPublic(c.rwc, c.EK.Handle())
+	if err != nil {
+		return nil, fmt.Errorf("cannot read EK from handle: %w", err)
+	}
+
+	encodedPublicEK, err := publicEK.Encode()
+	if err != nil {
+		return nil, fmt.Errorf("encode failed: %w", err)
+	}
+
+	return encodedPublicEK, nil
+}
+
+func (c *Context) createPolicySession() (tpmutil.Handle, error) {
+	var nonceCaller [32]byte
+	hSession, _, err := tpm2.StartAuthSession(
+		c.rwc,
+		tpm2.HandleNull,
+		tpm2.HandleNull,
+		nonceCaller[:],
+		nil,
+		tpm2.SessionPolicy,
+		tpm2.AlgNull,
+		tpm2.AlgSHA256,
+	)
+	if err != nil {
+		return 0, err
+	}
+
+	_, err = tpm2.PolicySecret(
+		c.rwc,
+		tpm2.HandleEndorsement,
+		tpm2.AuthCommand{Session: tpm2.HandlePasswordSession},
```

Comment:
is the TPM expecting a password to begin this session? If so, where is it defined?

Reply:
Unfortunately, this is not very well documented but my understanding is that the password that is expected here is the same password that was used to create the endorsement key.  Since the function `tpm2tools.EndorsementKeyRSA(...)` creates the endorsement key with an empty password, we leave the `Auth` field of the `AuthCommand` struct empty.